### PR TITLE
use atomics for expect/dbg

### PR DESCRIPTION
**Pull request by folkertdev**
the child is now locked while the parent prints the expect/dbg. That means that the printing happens at the right moment (e.g. in relation to the roc app printing to stdout itself with effects) and we can do multiple prints in sequence.

### CODE REVIEW

**Comment by folkertdev** [comment]
hmm, globals are used here but there is no `.data.rel.ro` section (#3609 is the issue we have for that). Here is the obj file that roc generates from the app

```
roc_app_bytes.o:     file format elf64-x86-64
roc_app_bytes.o
architecture: i386:x86-64, flags 0x00000011:
HAS_RELOC, HAS_SYMS
start address 0x0000000000000000

Sections:
Idx Name          Size      VMA               LMA               File off  Algn
  0 .text         000038d8  0000000000000000  0000000000000000  00000040  2**4
                  CONTENTS, ALLOC, LOAD, RELOC, READONLY, CODE
  1 .rodata.cst16 00000020  0000000000000000  0000000000000000  00003920  2**4
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  2 .rodata       00000040  0000000000000000  0000000000000000  00003940  2**3
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  3 .bss          00000008  0000000000000000  0000000000000000  00003980  2**3
                  ALLOC
  4 .note.GNU-stack 00000000  0000000000000000  0000000000000000  00003980  2**0
                  CONTENTS, READONLY
  5 .eh_frame     000001e0  0000000000000000  0000000000000000  00003980  2**3
                  CONTENTS, ALLOC, LOAD, RELOC, READONLY, DATA
SYMBOL TABLE:
0000000000000000 l    df *ABS*	0000000000000000 builtins-host
0000000000000000 l    d  .text	0000000000000000 .text
0000000000000000 l     F .text	00000000000000b2 roc_builtins.str.clone_to
0000000000000000 l       .rodata.cst16	0000000000000000 .LCPI1_0
0000000000000010 l       .rodata.cst16	0000000000000000 .LCPI1_1
00000000000000c0 l     F .text	000000000000070a roc_builtins.utils.expect_failed_start_shared_file
0000000000000018 l     O .rodata	0000000000000028 __unnamed_1
00000000000009f0 l     F .text	000000000000284d std.fmt.formatBuf
0000000000000000 l     O .rodata	0000000000000017 __unnamed_2
0000000000000000 l     O .bss	0000000000000008 SHARED_BUFFER.0
00000000000007d0 l     F .text	000000000000002d roc_builtins.utils.send_dbg
0000000000003240 l     F .text	0000000000000192 std.io.writer.Writer(*std.io.fixed_buffer_stream.FixedBufferStream([]u8),std.io.fixed_buffer_stream.WriteError,std.io.fixed_buffer_stream.FixedBufferStream([]u8).write).writeByteNTimes
00000000000033e0 l     F .text	0000000000000005 _mainForHost_67abdd721024f0ff4eb3f4c2fc13bc5bad42db7851d456d88d203d15aaa450
00000000000034c0 l     F .text	0000000000000207 #UserApp_main_df3f61984a92fdb4057192dc43dd748ea778adc52bc498ce8524c014b81119
0000000000003460 l     F .text	0000000000000001 Task_53_dc765660b06ee03dd16fd7ca5b957e8c805161ac2c4af28c5a10ab2ab432ca1
0000000000003470 l     F .text	0000000000000022 Task_putLine_fb5e512425fc9449316ec95969ebe71e2d576dbab833d61e2a5b9330fd70ee2
00000000000036d0 l     F .text	000000000000003d Effect_putLine_2594b6a92ebfb1c3312deb7d1c015fb95e9fbe9bd7bc6b527af7813ec7b910
00000000000034b0 l     F .text	000000000000000c Effect_map_9d9f29527a6be626a8f5985b26e19b237b44872b03631811df4416fc1713178
00000000000034a0 l     F .text	0000000000000005 _2_26b25d457597a7b0463f9620f666dd10aa2c4373a55967c7c8d70922a2d6ece
0000000000003710 l     F .text	000000000000005c Effect_effect_map_inner_7aa8ca4a2506da9133d8f889678b76f716ce45d2e22fdb7b70a15e56a0eff8
0000000000003770 l     F .text	0000000000000015 Effect_effect_closure_putLine_e8613f5a5bc9f9feeda32a8e7c80b69dd4878e47b6a91723fb15eb84236b6a2b
00000000000038d0 l     F .text	0000000000000008 roc_fx_putLine_fastcc_wrapper
0000000000003800 l     F .text	000000000000004e decrement_refcounted_ptr_8
0000000000000000 l    d  .rodata	0000000000000000 .rodata
0000000000000000 l    d  .bss	0000000000000000 .bss
0000000000000000         *UND*	0000000000000000 memcpy
0000000000000000         *UND*	0000000000000000 roc_getppid
0000000000000000         *UND*	0000000000000000 roc_shm_open
0000000000000000         *UND*	0000000000000000 roc_mmap
0000000000000800  w    F .text	00000000000001e6 __muloti4
0000000000000000         *UND*	0000000000000000 __udivti3
0000000000000000         *UND*	0000000000000000 memset
00000000000033f0 g     F .text	000000000000001b roc__mainForHost_1_exposed_generic
0000000000003410 g     F .text	000000000000003b roc__mainForHost_1_exposed
0000000000003450 g     F .text	0000000000000007 roc__mainForHost_size
0000000000003790 g     F .text	0000000000000013 roc__mainForHost_1__Fx_caller
00000000000037b0 g     F .text	0000000000000003 roc__mainForHost_1__Fx_result_size
00000000000037c0 g     F .text	0000000000000007 roc__mainForHost_1__Fx_size
0000000000000000         *UND*	0000000000000000 roc_dealloc
0000000000000000         *UND*	0000000000000000 roc_fx_putLine


RELOCATION RECORDS FOR [.text]:
OFFSET           TYPE              VALUE 
000000000000009a R_X86_64_PLT32    memcpy-0x0000000000000004
00000000000000d8 R_X86_64_PLT32    roc_getppid-0x0000000000000004
000000000000012e R_X86_64_PC32     .LCPI1_0-0x0000000000000004
0000000000000145 R_X86_64_PC32     .LCPI1_1-0x0000000000000004
0000000000000298 R_X86_64_PC32     .rodata+0x0000000000000014
00000000000003c8 R_X86_64_PC32     .rodata-0x0000000000000004
00000000000003dc R_X86_64_PC32     .rodata-0x0000000000000004
00000000000004cc R_X86_64_PC32     .rodata-0x0000000000000004
00000000000004da R_X86_64_PC32     .rodata-0x0000000000000004
0000000000000515 R_X86_64_PC32     .rodata-0x0000000000000004
0000000000000523 R_X86_64_PC32     .rodata-0x0000000000000004
00000000000005bc R_X86_64_PC32     .rodata-0x0000000000000004
00000000000005c9 R_X86_64_PC32     .rodata-0x0000000000000004
00000000000006a0 R_X86_64_PC32     .rodata-0x0000000000000004
000000000000071d R_X86_64_PC32     .rodata-0x0000000000000004
0000000000000799 R_X86_64_PLT32    roc_shm_open-0x0000000000000004
00000000000007b7 R_X86_64_PLT32    roc_mmap-0x0000000000000004
00000000000007be R_X86_64_PC32     .bss-0x0000000000000004
00000000000007d3 R_X86_64_PC32     .bss-0x0000000000000004
0000000000000989 R_X86_64_PLT32    __udivti3-0x0000000000000004
00000000000009c4 R_X86_64_PLT32    __udivti3-0x0000000000000004
0000000000001b79 R_X86_64_PLT32    memset-0x0000000000000004
0000000000001cad R_X86_64_PLT32    memcpy-0x0000000000000004
0000000000001d7d R_X86_64_PLT32    memset-0x0000000000000004
0000000000001ed7 R_X86_64_PLT32    memcpy-0x0000000000000004
0000000000002562 R_X86_64_PLT32    memset-0x0000000000000004
0000000000002696 R_X86_64_PLT32    memcpy-0x0000000000000004
0000000000003262 R_X86_64_PLT32    memset-0x0000000000000004
0000000000003365 R_X86_64_PLT32    memcpy-0x0000000000000004
0000000000003845 R_X86_64_PLT32    roc_dealloc-0x0000000000000004
00000000000038d2 R_X86_64_PLT32    roc_fx_putLine-0x0000000000000004


RELOCATION RECORDS FOR [.eh_frame]:
OFFSET           TYPE              VALUE 
0000000000000020 R_X86_64_PC32     .text+0x00000000000033e0
0000000000000034 R_X86_64_PC32     .text+0x00000000000033f0
000000000000004c R_X86_64_PC32     .text+0x0000000000003410
0000000000000064 R_X86_64_PC32     .text+0x0000000000003450
0000000000000078 R_X86_64_PC32     .text+0x0000000000003460
000000000000008c R_X86_64_PC32     .text+0x0000000000003470
00000000000000a4 R_X86_64_PC32     .text+0x00000000000034a0
00000000000000b8 R_X86_64_PC32     .text+0x00000000000034b0
00000000000000cc R_X86_64_PC32     .text+0x00000000000034c0
00000000000000e8 R_X86_64_PC32     .text+0x00000000000036d0
0000000000000100 R_X86_64_PC32     .text+0x0000000000003710
0000000000000118 R_X86_64_PC32     .text+0x0000000000003770
0000000000000130 R_X86_64_PC32     .text+0x0000000000003790
0000000000000148 R_X86_64_PC32     .text+0x00000000000037b0
000000000000015c R_X86_64_PC32     .text+0x00000000000037c0
0000000000000170 R_X86_64_PC32     .text+0x00000000000037d0
000000000000018c R_X86_64_PC32     .text+0x0000000000003800
00000000000001a4 R_X86_64_PC32     .text+0x0000000000003850
00000000000001b8 R_X86_64_PC32     .text+0x00000000000038b0
00000000000001d0 R_X86_64_PC32     .text+0x00000000000038d0
```

The `rodata.cst16` looks a little odd. Maybe we need to handle it in a custom way (rather than lumping it in with the rest of rodata?)

**Comment by folkertdev** [comment]
what did you try? CI failures seem to be something with the repl. @rtfeldman confirmed that `roc dev` does now work on macos (specifically in release builds)

**Comment by rtfeldman** [comment]
That said, the repl tests do fail for me on M1, just like on CI. 🤔

**Comment by rtfeldman** [comment]
Hm, but the repl tests also fail for me on `main` on M1 😅

So it might not be this PR!

**Comment by rtfeldman** [comment]
Yeah I just realized that too. 🤦 

Re-running things now!

**Comment by folkertdev** [comment]
@bhansconnect quick guess: the alignment of ` .rodata.cst16` is 16, normal `.rodata` is only 8

**Comment by bhansconnect** [comment]
> the alignment of .rodata.cst16 is 16, normal .rodata is only 8

Definitely could be the issue, but I thought we were already dealing with this. We align all sections to 64.

**Comment by folkertdev** [comment]
yes, I played around with it and that was not the issue. I've also compared `.o` files, and the only real difference seems to be additional data in the `.bss` section (that constant is zero-initialized). 

Anyway, I found a way around the global for this PR that is actually much nicer anyway. So we can kick this linker problem can down the road

File: crates/repl_expect/src/run.rs [hunk 1]
```diff
@@ -611,11 +629,47 @@ impl ExpectSequence {
...
+        let ptr = self.ptr as *const u32;
+        let atomic_ptr: *const AtomicU32 = unsafe { ptr.add(5).cast() };
+        let atomic = unsafe { &*atomic_ptr };
+
+        loop {
+            if sigchld.load(Ordering::Relaxed) {
+                break ChildProcessMsg::Terminate;
+            }
+
+            match atomic.load(Ordering::Acquire) {
```
Comment (ayazhafiz) [inline crates/repl_expect/src/run.rs:644]: Is there a reason a relaxed ordering can't be used here as well?

**Review by ayazhafiz — APPROVED** [review]

File: crates/repl_expect/src/run.rs (hunk 1 above)
Comment (folkertdev) [reply crates/repl_expect/src/run.rs:644]: no, I'll change that in the followup. In general the atomics might not even be strictly needed here, there is never contention

**Pull request merged**